### PR TITLE
CHE-7098: Add an option to ask user before browser tab closing

### PR DESCRIPTION
**Pull request by mmorhun**
### What does this PR do?
Adds an ability to prevent closing of Che IDE tab. This behaviour can be set in user preferences. By default this option is turned off.
As before, check for unsaved changes in editor is performed despite the option configuration. Then, if no unsaved changes, the option take place.
In case of programmatic IDE leaves (GWT dev mode) all checks are ignored.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/7098

#### Release Notes
Adds an option to prevent closing of Che IDE tab.

### CODE REVIEW

**Comment by mmorhun** [comment]
ci-build

**Comment by mmorhun** [comment]
ci-test

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/CoreLocalizationConstant.java [hunk 1]
```diff
@@ -50,11 +50,11 @@
   @Key("navigateToFile.searchIsCaseSensitive")
   String navigateToFileSearchIsCaseSensitive();
 
-  @Key("appearance.title")
-  String appearanceTitle();
+  @Key("general.title")
+  String generalTitle();
```
Comment (tolusha) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/CoreLocalizationConstant.java]: What was the reason to refactor the code?

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/BrowserTabCloseHandler.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    this.preferencesManager = preferencesManager;
+
+    eventBus.addHandler(WindowActionEvent.TYPE, this);
+  }
+
+  @Override
+  public void onWindowClosing(WindowActionEvent event) {
+    if (!closeImmediately) {
+      if (!isEditorsClean()) {
+        event.setMessage("ask user");
```
Comment (tolusha) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/BrowserTabCloseHandler.java]: It looks like user will see the 'ask user' message.

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/preferences/pages/general/IdeGeneralPreferencesPresenter.java [hunk 3]
```diff
@@ -20,67 +20,86 @@
...
       ThemeAgent themeAgent,
       PreferencesManager preferencesManager) {
-    super(constant.appearanceTitle(), constant.appearanceCategory());
+    super(constant.generalTitle(), constant.generalCategory());
     this.view = view;
     this.themeAgent = themeAgent;
     this.preferencesManager = preferencesManager;
     view.setDelegate(this);
+
+    revertChanges();
```
Comment (tolusha) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/preferences/pages/general/IdeGeneralPreferencesPresenter.java]: It looks weird.

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/BrowserTabCloseHandler.java (hunk 2 above)
Comment (mmorhun) [reply ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/BrowserTabCloseHandler.java]: No, actually their won't. Modern browsers block custom messages, so now matters only whether it was null or a string. See `setMessage` description for more details.

Comment (tolusha) [reply ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/BrowserTabCloseHandler.java]: If browser starts taking into account messages then we will have to alter the message.
So, it is better to set the message if it is really displayed.

Comment (mmorhun) [reply ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/BrowserTabCloseHandler.java]: It won't. At least the probability is too small (they deleted such ability on purpose).
But ok, I will add messages...

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/preferences/pages/general/IdeGeneralPreferencesPresenter.java (hunk 3 above)
Comment (mmorhun) [reply ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/preferences/pages/general/IdeGeneralPreferencesPresenter.java]: ok, I'll rename method

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/preferences/pages/general/IdeGeneralPreferencesViewImpl.java [hunk 4]
```diff
@@ -56,11 +59,26 @@ public void setThemes(List<Theme> themes, String currentThemeId) {
...
     delegate.themeSelected(themeBox.getValue(themeBox.getSelectedIndex()));
   }
 
-  interface AppearanceViewImplUiBinder extends UiBinder<FlowPanel, AppearanceViewImpl> {}
+  @UiHandler("askBeforeClosingTab")
+  void onAskBeforeClosingTabChanged(final ValueChangeEvent<Boolean> event) {
+    delegate.onAskBeforeClosingTabChanged(event.getValue());
+  }
+
+  interface AppearanceViewImplUiBinder extends UiBinder<FlowPanel, IdeGeneralPreferencesViewImpl> {}
```
Comment (azatsarynnyy) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/preferences/pages/general/IdeGeneralPreferencesViewImpl.java]: rename `AppearanceViewImplUiBinder` --> `IdeGeneralPreferencesUiBinder` please

**Review by azatsarynnyy — APPROVED** [review]

**Review by tolusha — APPROVED** [review]

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/preferences/pages/general/IdeGeneralPreferencesViewImpl.java (hunk 4 above)
Comment (mmorhun) [reply ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/preferences/pages/general/IdeGeneralPreferencesViewImpl.java]: thank you, missed it

**Review by ghost — APPROVED** [review]

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/BrowserTabCloseHandler.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+   * @return true if no unsaved changes, false otherwise
+   */
+  private boolean isEditorsClean() {
+    for (EditorPartPresenter editorPartPresenter : editorAgentProvider.get().getOpenedEditors()) {
+      if (editorPartPresenter.isDirty()) {
+        return false;
+      }
+    }
+
+    return true;
```
Comment (vinokurig) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/BrowserTabCloseHandler.java]: `return editorAgentProvider.get().getOpenedEditors().stream().noneMatch(editor -> editor.isDirty())` ?

Comment (mmorhun) [reply ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/BrowserTabCloseHandler.java]: This is shorter, ok, let be so.

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/preferences/pages/general/IdeGeneralPreferencesViewImpl.ui.xml [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+            margin-right: 5px;
+        }
+    </ui:style>
+    <g:FlowPanel styleName="{style.main}">
+        <g:FlowPanel addStyleNames="{style.row}">
+            <g:Label addStyleNames="{style.leftAlign}">Theme</g:Label>
+            <che:CustomListBox ui:field="themeBox" debugId="window-preferences-general-themeBox" addStyleNames="{style.rightAlign} {style.dropDownShift}"/>
+        </g:FlowPanel>
+        <g:FlowPanel addStyleNames="{style.row}">
+            <g:Label addStyleNames="{style.leftAlign}">Ask before closing browser tab</g:Label>
```
Comment (vinokurig) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/preferences/pages/general/IdeGeneralPreferencesViewImpl.ui.xml]: Please move text to locales

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/preferences/pages/general/IdeGeneralPreferencesViewImpl.ui.xml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+            height: 25px;
+            margin-top: 5px;
+        }
+        .dropDownShift {
+            margin-right: 5px;
+        }
+    </ui:style>
+    <g:FlowPanel styleName="{style.main}">
+        <g:FlowPanel addStyleNames="{style.row}">
+            <g:Label addStyleNames="{style.leftAlign}">Theme</g:Label>
```
Comment (vinokurig) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/preferences/pages/general/IdeGeneralPreferencesViewImpl.ui.xml]: Please move text to locales

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/preferences/pages/general/IdeGeneralPreferencesViewImpl.ui.xml (hunk 6 above)
Comment (mmorhun) [reply ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/preferences/pages/general/IdeGeneralPreferencesViewImpl.ui.xml]: ok

**Review by vinokurig — APPROVED** [review]

**Review by vparfonov — APPROVED** [review]

**Pull request merged**